File: software/python/asm.py
"""16bit hack assembler.

Spec:

a instructions start with 0
c instructions start with 1

example c instruction:

c_inst = 1 1 1 a c1 c2 c3 c4 c5 c6 d1 d2 d3 j1 j2 j3
c_inst in text => dest = comp; jump

c_inst[0] = op code
c_inst[1 - 2] = not used
c_inst[3 - 9] = comp bits
c_inst[10 - 12] = dest bits
c_inst[13 - 15] = jump bits

comp bits table if a = 0:
0: 101010,
1: 111111,
-1: 111010,
"D": 001100,
"A": 110000,
"!D": 001101,
"!A": 110001,
"-D": 001111,
"-A": 110011,
"D+1": 011111,
"A+1": 110111,
"D-1": 001110,
"A-1": 110010,
"D+A": 000010,
"D-A": 010011,
"A-D": 000111,
"D&A": 000000,
"D|A": 010101,

comp bits table if a = 1:
"M": "A",
"!M": "!A",
"-M": "-A",
"M+1": "A+1",
"M-1": "A-1",
"D+M": "D+A",
"D-M": "D-A",
"M-D": "A-D",
"D&M": "D&A",
"D|M": "D|A",

dest bits table:
"null": 000,  # the value is not stored
"M": 001,     # RAM[A]
"D": 010,     # D register
"MD": 011,    # RAM[A] and D register
"A": 100,     # A register
"AM": 101,    # A register and RAM[A]
"AD": 110,    # A register and D register
"AMD": 111,   # A register, RAM[A], and D register

jump bits table:
"null": 000,  # no jump
"JGT": 001,   # if out > 0 jump
"JEQ": 010,   # if out == 0 jump
"JGE": 011,   # if out >= 0 jump
"JLT": 100,   # if our < 0 jump
"JNE": 101,   # if out != 0 jump
"JLE": 110,   # if out <= 0 jump
"JMP": 111,   # unconditional jump

symbol table for built in symbols:
"R0": 0,
"R1": 1,
"R2": 2,
"R3": 3,
"R4": 4,
"R5": 5,
"R6": 6,
"R7": 7,
"R8": 8,
"R9": 9,
"R10": 10,
"R11": 11,
"R12": 12,
"R13": 13,
"R14": 14,
"R15": 15,
"SCREEN": 16384,
"KBD": 24576,
"SP": 0,
"LCL": 1,
"ARG": 2,
"THIS": 3,
"THAT": 4,

"""
import argparse
import pathlib
import sys

import emit
import utils

# symbol tables
dest: dict[str, str] = {
    "null": "000",
    "M": "001",
    "D": "010",
    "MD": "011",
    "A": "100",
    "AM": "101",
    "AD": "110",
    "AMD": "111",
}

jump: dict[str, str] = {
    "null": "000",
    "JGT": "001",
    "JEQ": "010",
    "JGE": "011",
    "JLT": "100",
    "JNE": "101",
    "JLE": "110",
    "JMP": "111",
}

comp_a_is_0: dict[str, str] = {
    # when a == 0
    "0": "101010",
    "1": "111111",
    "-1": "111010",
    "D": "001100",
    "A": "110000",
    "!D": "001101",
    "!A": "110001",
    "-D": "001111",
    "-A": "110011",
    "D+1": "011111",
    "A+1": "110111",
    "D-1": "001110",
    "A-1": "110010",
    "D+A": "000010",
    "D-A": "010011",
    "A-D": "000111",
    "D&A": "000000",
    "D|A": "010101",
}


comp_a_is_1: dict[str, str] = {
    # when a == 1
    "M": "110000",
    "!M": "110001",
    "-M": "110011",
    "M+1": "110111",
    "M-1": "110010",
    "D+M": "000010",
    "D-M": "010011",
    "M-D": "000111",
    "D&M": "000000",
    "D|M": "010101",
}


symbols: dict[str, str] = {
    # registers on CPU
    "R0": "0",
    "R1": "1",
    "R2": "2",
    "R3": "3",
    "R4": "4",
    "R5": "5",
    "R6": "6",
    "R7": "7",
    "R8": "8",
    "R9": "9",
    "R10": "10",
    "R11": "11",
    "R12": "12",
    "R13": "13",
    "R14": "14",
    "R15": "15",

    # extra variables used by jack VM
    "SP": "0",
    "LCL": "1",
    "ARG": "2",
    "THIS": "3",
    "THAT": "4",

    # peripherals
    "SCREEN": "16384",
    "KBD": "24576",
}


globals = {
    "next_register": 16,
    # holds the count of valid lines of code i.e. not whitespace, comments etc
    "line_count": 0,
}


def is_comment(line, idx, len):
    return (idx + 1 < len) and (line[idx + 1] == "/")


def get_variable(line: str) -> str:
    """Parse an asm variable.

    According to the hack asm spec, variables start with the @ symbol.
    """
    if not line.startswith("@"):
        raise ValueError("Invalid Variable")

    var: list[str] = []

    for i in range(len(line)):
        char: str = line[i]

        if char == "@":
            continue

        # reached end of variable
        if char == " " or char == "\n":
            break

        # reached a comment
        if is_comment(line, i, len(line)):
            break

        # char is part of variable
        var.append(char)

    return "".join(var)


def get_symbol(line: str) -> str:
    """Parse an asm symbol.

    According to hack spec, all valid symbols start with '('
    and end with ')' and contain no spaces.
    """
    if not line.startswith("("):
        raise ValueError("Invalid variable definition")

    var: list[str] = []
    closed: bool = False

    for i in range(len(line)):
        char = line[i]

        if char == "(":
            continue

        # reached end of variable
        if char == ")":
            closed = True
            break

        # contains spaces and is invalid
        if char == " ":
            break

        # reached a comment
        if is_comment(line, i, len(line)):
            break

        # char is part of variable
        var.append(char)

    if not closed:
        raise ValueError("Invalid variable definition")

    return "".join(var)


def _parse_instruction(line):
    if not line:
        raise ValueError("Invalid instruction")

    # separate components of the instruction
    s = line.split("=")
    s2 = s[1].split(";") if len(s) == 2 else s[0].split(";")

    out = {
        "dest": s[0] if len(s) == 2 else "null",
        "comp": s2[0],
        "jump": s2[1] if len(s2) == 2 else "null",
    }

    return out


def instruction(line):
    out = _parse_instruction(line)

    x = ["1","1","1"]
    # byte order is comp, dest, jump
    if out["comp"] in comp_a_is_0:
        x.append("0")
        x.append(comp_a_is_0[out["comp"]])
    else:
        x.append("1")
        x.append(comp_a_is_1[out["comp"]])

    x.append(dest[out["dest"]])
    x.append(jump[out["jump"]])

    return "".join(x)


def variable(line):
    var = get_variable(line)

    # we want to convert the variable to a number
    # where this number comes from depends on whether the
    # variable is aleady a decimal number or if we need to
    # use the next available register as a proxy
    as_num = None

    if utils.is_number(var):
        as_num = utils.atoi(var)

    else:
        # look up in symbol table
        if var not in symbols:
            # convert to string to maintain consistency with the rest of our
            # symbol mapping
            symbols[var] = str(globals["next_register"])
            # increment the register after using
            globals["next_register"] += 1

        as_num = utils.atoi(symbols[var])

    # turn the decimal number to binary
    dec_bin = utils.dec_to_bin(as_num)

    return dec_bin


def assemble(fp):
    # build symbole table in a first pass
    build_symbol_table(fp)

    for line in utils.readlines(fp):
        fist_char_idx = utils.first(line)

        # -1 means its an empty str
        if fist_char_idx == -1: continue
        # skip labels as we've already accounted for them during symbol table
        # building first pass
        if line[fist_char_idx] == "(": continue

        # valid line of hack asm
        cleaned = utils.clean(line)

        # hadle variable
        if cleaned.startswith("@"):
            emit.emit(variable(cleaned))
        # handle instruction
        else:
            emit.emit(instruction(cleaned))


    # flush anything left in the buffer
    emit.flush()


def build_symbol_table(fp):
    for line in utils.readlines(fp):
        if utils.first(line) == -1: continue
        # valid line of hack asm
        line = utils.clean(line)

        if line.startswith("("):
            symbol = get_symbol(line)
            # set symbol to point to the next line
            # essentially ignore this symbol in our line count
            symbols[symbol] = str(globals["line_count"])

        else:
            globals["line_count"] += 1


def main():
    """Command line tool for assembling hack ASM to machine code."""

    parser = argparse.ArgumentParser(
        prog="Assemble", description="N2T Hack Assembler")

    parser.add_argument(
        "file",
        type=pathlib.Path,
        help="Path to file containing hack. Note: You must have read "
             "privileges on said file."
    )

    parser.add_argument(
        "--out",
        type=pathlib.Path,
        default="out.hack",
        help="file to write output to."
    )

    parser.add_argument(
        "--symbols-only",
        action="store_true",
        help="Print the symbol table to the console."
    )

    return parser.parse_args()


if __name__ == "__main__":
    args = main()

    if not args.file.exists():
        print("FILE DOES NOT EXIST!!")
        sys.exit(1)

    if args.symbols_only:
        build_symbol_table(args.file)
        print(symbols)

    else:
        # initialize .hack file
        emit.init(args.out)

        assemble(args.file)
        print("Successfully assembled!!")

    sys.exit(0)

File: software/c/asm.c
#include <stdio.h>
#include <stdlib.h>

#include "array.h"
#include "mystring.h"
#include "hash.h"


// ------------------ prototypes ------------
void push_callback(Item *, void *);

// ----------- ---- UTILS -----------------

int first(String str)
{
    int result = -1;

    for (size_t i = 0; i < str.length; i++) {
        char curr = charat(str, i);

        // space character, we can ignore
        if (curr == ' ')
            continue;

        // empty line
        if (curr == '\n')
            break;

        // comment
        if (
            curr == '/' &&
            (i + 1 < str.length && str.s[i + 1] == '/')
        )
            break;

        // we've found the first proper char
        result = (int)i;
        break;
    }

    return result;
}


String clean(String str)
{
    String ns = newstr(""); 
    size_t i = 0;

    while (i < str.length && str.s[i] == ' ')
        i++;

    while (i < str.length) {
        char curr = charat(str, i);

        // reached the end of valid asm
        // NOTE: this does not support space inbetween valid asm statements
        if (curr == ' ' || curr == '\n')
            break;

        // comment
        if (
            curr == '/' &&
            (i + 1 < str.length && str.s[i + 1] == '/')
        )
            break;

        ns = concatChar(ns, &curr);
        i++;
    }

    return ns;
}


String str_reverse(String str)
{
    String ns = newstr("");

    // cast to int as size_t doesn't go below 0
    int i = (int)str.length - 1;

    while (i >= 0) {
        char curr = charat(str, (size_t)i);
        ns = concatChar(ns, &curr);

        i--;
    }

    return ns;
}


String dec_to_str(unsigned int num)
{
    String ns = newstr("");

    char d[2];
    d[1] = '\0';

    while (num > 0) {
        char tmp = '0' + (num % 10);
        num /= 10;

        d[0] = tmp;
        ns = concatChar(ns, d);

    }

    return str_reverse(ns);
}


bool is_number(String str)
{
    if (str.s == NULL)
        return false;

    // check if number is signed
    size_t start_idx = (str.s[0] == '-') ? 1 : 0;

    for (size_t i = start_idx; i < str.length; i++) {
        char curr = charat(str, i);

        if (!(('0' <= curr) && (curr <= '9')))
            return false;
    }

    return true;
}


long myatoi(String str)
{
    // check if number is signed
    int sign = (str.s[0] == '-') ? -1 : 1;
    size_t start_idx = (sign == -1) ? 1 : 0;

    long result = 0;

    for (size_t i = start_idx; i < str.length; i++) {
        char curr = charat(str, i);

        result *= 10;
        result += (curr - '0');
    }

    return result * sign;
}


String dec_to_bin(long number)
{
    String result = newstr("");

    while (number >= 1) {
        // modding by 2 will give us either 0 or 1
        char mod = '0' + (number % 2);
        result = concatChar(result, &mod);
        // divide by two to move down to the next power of 2
        number /= 2;
    }

    // pad any space with 0s because we want 16bit values
    while ((result = concatChar(result, "0")).length < 16);

    return str_reverse(result);
}


Array * str_split(String str, char *symbol)
{
    Array *arr = newarr(push_callback);

    // ensure we have only one character to compare to
    char needle = symbol[0];

    // create a new string containing all letters up to the symbol
    String tmp = newstr("");

    for (size_t i = 0; i < str.length; i++) {
        char curr = charat(str, i);

        if (curr == needle) {

            if (tmp.length) {
                // we have a valid string
                push(arr, &tmp);
                // reset string
                tmp = newstr("");
            }

            continue;

        }

        tmp = concatChar(tmp, &curr);
    }

    // check if there is a string left over
    if (tmp.length)
        push(arr, &tmp);

    return arr;
}


void exit_with_messages(char *message)
{
    printf("%s\n", message);
    exit(1);
}


// ------------ GLOBAL CONSTANTS ----------

// holds the count of valid lines of code i.e. not whitespace, comments etc
size_t LINE_CNT = 0;
// next available register, starts at 16 as hack uses first 15 by default
unsigned int NXT_REG = 16;

// initialized at run time
static HT *DEST = NULL;     // op codes for destination of hack asm
static HT *JUMP = NULL;     // op codes for jump related asm
static HT *COMP_0 = NULL;   // op codes for when instruction starts with 0
static HT *COMP_1 = NULL;   // op codes for when instruction starts with 1
static HT *SYMBOLS = NULL;  // user defined symbols/variables

// ------------- Symbol tables ------------
// to make things easier we hard code the basic asm -> machine code symbols


HT * dest(void)
{
    HT *map = create();

    set(map, "null", "000");
    set(map, "M", "001");
    set(map, "D", "010");
    set(map, "MD", "011");
    set(map, "A", "100");
    set(map, "AM", "101");
    set(map, "AD", "110");
    set(map, "AMD", "111");

    return map;
}


HT * jump(void)
{
    HT *map = create();

    set(map, "null", "000");
    set(map, "JGT", "001");
    set(map, "JEQ", "010");
    set(map, "JGE", "011");
    set(map, "JLT", "100");
    set(map, "JNE", "101");
    set(map, "JLE", "110");
    set(map, "JMP", "111");

    return map;
}


HT * comp_zero(void)
{
    HT *map = create();

    set(map, "0", "101010");
    set(map, "1", "111111");
    set(map, "-1", "111010");
    set(map, "D", "001100");
    set(map, "A", "110000");
    set(map, "!D", "001101");
    set(map, "!A", "110001");
    set(map, "-D", "001111");
    set(map, "-A", "110011");
    set(map, "D+1", "011111");
    set(map, "A+1", "110111");
    set(map, "D-1", "001110");
    set(map, "A-1", "110010");
    set(map, "D+A", "000010");
    set(map, "D-A", "010011");
    set(map, "A-D", "000111");
    set(map, "D&A", "000000");
    set(map, "D|A", "010101");

    return map;
}


HT * comp_one(void)
{
    HT *map = create();

    set(map, "M", "110000");
    set(map, "!M", "110001");
    set(map, "-M", "110011");
    set(map, "M+1", "110111");
    set(map, "M-1", "110010");
    set(map, "D+M", "000010");
    set(map, "D-M", "010011");
    set(map, "M-D", "000111");
    set(map, "D&M", "000000");
    set(map, "D|M", "010101");

    return map;
}


HT * symbols(void)
{
    HT *map = create();

    // base registers on CPU
    set(map, "R0", "0");
    set(map, "R1", "1");
    set(map, "R2", "2");
    set(map, "R3", "3");
    set(map, "R4", "4");
    set(map, "R5", "5");
    set(map, "R6", "6");
    set(map, "R7", "7");
    set(map, "R8", "8");
    set(map, "R9", "9");
    set(map, "R10", "10");
    set(map, "R11", "11");
    set(map, "R12", "12");
    set(map, "R13", "13");
    set(map, "R14", "14");
    set(map, "R15", "15");

    // variables use by the jack VM
    set(map, "SP", "0");
    set(map, "LCL", "1");
    set(map, "ARG", "2");
    set(map, "THIS", "3");
    set(map, "THAT", "4");

    // peripherals
    set(map, "SCREEN", "16384");
    set(map, "KBD", "24576");

    return map;
}


void push_callback(Item *it, void *data)
{
    String *ndata = calloc(sizeof(ndata), 1);
    *ndata = *(String *)data;
    it->data = ndata;
}


Array * readlines(char *path)
{
    int c;
    // depending on how the file is saved e.g. as UTF-8, we can have
    // problems reading characters. For example, a byte order mark may be
    // automatically added to the first line, so characters come out as
    // gibberish. Rather than perfectly trying to decode any unicode character,
    // here we simpply make a one letter null terminated 'string' but casting
    // the incoming character and assigning its value to the first index of
    // this array.
    char d[2];

    Array *arr = newarr(push_callback);

    FILE *fp = fopen(path, "rb");

    String line = newstr("");

    d[1] = '\0';

    while ((c = fgetc(fp)) != EOF) {
        if (c == '\n') {

            push(arr, &line);
            line = newstr("");
            continue;
        }

        char c2 = (char)c;
        d[0] = c2;

        line = concatChar(line, d);
    }

    fclose(fp);

    return arr;

}


void write_hack(String bin)
{
    if (bin.length == 0)
        return;

    FILE *fp = fopen("out.hack", "a");

    fprintf(fp, "%s\n", bin.s);

    // cleanup
    fclose(fp);
    freestr(bin);
}


String get_symbol(String str)
{
    int closed = 0;
    String ns = newstr("");

    for (size_t i = 0; i < str.length; i++) {
        char curr = charat(str, i);

        if (curr == '(')
            continue;

        // end of variable
        // Note: this does not support spaces around brackets
        if (curr == ' ' || curr == '\n')
            break;

        // comment
        if (
            curr == '/' &&
            (i + 1 < str.length && str.s[i + 1] == '/')
        )
            break;

        if (curr == ')') {
            closed = 1;
            break;
        }

        ns = concatChar(ns, &curr);
    }

    if (!closed)
        exit_with_messages("Invalid variable definition");

    return ns;
}


String get_variable(String str)
{
    String ns = newstr("");

    for (size_t i = 0; i < str.length; i++) {
        char curr = charat(str, i);

        if (curr == '@')
            continue;

        // end of variable
        // Note: this does not support spaces between '@' and var name
        if (curr == ' ' || curr == '\n')
            break;

        // comment
        if (
            curr == '/' &&
            (i + 1 < str.length && str.s[i + 1] == '/')
        )
            break;

        ns = concatChar(ns, &curr);
    }

    return ns;
}


void build_symbol_table(Array *arr)
{
    for (size_t i = 0; i < arr->length; i++) {
        String line = *(String *)arr->items[i].data;

        if (first(line) == -1)
            continue;

        line = clean(line);

        if (startswith(line, "(")) {
            String symbol = get_symbol(line);
            const char *result = set(SYMBOLS, symbol.s, dec_to_str(LINE_CNT).s);

            if (result == NULL)
                printf("Did not save: %s", symbol.s);

        } else {
            LINE_CNT++;
        }

    }
}


HT * _instruction_parser(String line)
{
    Array *s;
    Array *s2;

    HT *map = create();

    // init values
    set(map, "dest", "null");
    set(map, "jump", "null");
    set(map, "comp", NULL);

    // Both jump and dest are optional but at least one of them has to be present
    // so this is a bit of a pain but necessary
    // Future changes should make this more robust to asm syntax errors.
    s = str_split(line, "=");
    s2 = str_split(*(String *)s->items[s->length == 2 ? 1 : 0].data, ";");

    // set comp
    String tmp = *(String *)s2->items[0].data;
    set(map, "comp", tmp.s);

    // set dest if present
    if (s->length == 2) {
        String curr = *(String *)s->items[0].data;
        set(map, "dest", curr.s);
    }

    // set jump if present
    if (s2->length == 2) {
        String curr = *(String *)s2->items[1].data;
        set(map, "jump", curr.s);
    }

    // cleanup
    freearray(s);
    freearray(s2);

    return map;
}


String instruction_parse(String line)
{
    HT *parsed = _instruction_parser(line);

    // holds the binary result of the instruction
    String result = newstr("111");

    char *comp = (char *)get(parsed, "comp");
    char *dst = (char *)get(parsed, "dest");
    char *jmp = (char *)get(parsed, "jump");

    if (!comp)
        printf("COMP NOT FOUND!\n");

    if (get(COMP_0, comp) != NULL) {
        result = concatChar(result, "0");
        result = concatRaw(result, (char *)get(COMP_0, comp));

    } else {
        result = concatChar(result, "1");
        result = concatRaw(result, (char *)get(COMP_1, comp));
    }

    result = concatRaw(result, (char *)get(DEST, dst));
    result = concatRaw(result, (char *)get(JUMP, jmp));

    // cleanup
    destroy(parsed);

    return result;
}


String variable(String str)
{
    long as_num;

    String var = get_variable(str);

    if (is_number(str)) {
        as_num = myatoi(var);

    } else {
        // look up in symbol table
        if (get(SYMBOLS, var.s) == NULL) {
            // convert to string to maintain consistency with the rest of our
            // symbol mapping
            set(SYMBOLS, var.s, dec_to_str(NXT_REG).s);
            // increment the register after using
            NXT_REG++;
        }

        char *sym = get(SYMBOLS, var.s);
        as_num = myatoi(newstr(sym));
    }

    String dec_bin = dec_to_bin(as_num);

    return dec_bin;
}


void assemble(Array *arr)
{
    // build the symbol table for user declared variables
    build_symbol_table(arr);

    for (size_t i = 0; i < arr->length; i++) {
        String line = *(String *)arr->items[i].data;

        int first_char_idx = first(line);

        if (first_char_idx == -1) continue;
        if (line.s[first_char_idx] == '(') continue;

        String cleaned = clean(line);

        if (startswith(cleaned, "@"))
            write_hack(variable(cleaned));
        else
            write_hack(instruction_parse(cleaned));
    }
}


void init_tables()
{
    DEST = dest();
    JUMP = jump();
    COMP_0 = comp_zero();
    COMP_1 = comp_one();
    SYMBOLS = symbols();
}


int main()
{
    // init symbol tables    
    init_tables();

    Array *arr = readlines("t.asm");
    assemble(arr);
    return 0;
}
